### PR TITLE
ADD Auto test

### PR DESCRIPTION
**Pull request by pushtheworldllc**

### CODE REVIEW

File: openBCIBoard.js [hunk 1]
```diff
@@ -75,60 +94,78 @@ function OpenBCIFactory() {
...
                     if(this.options.verbose) console.log("Waiting for '$$$'");
 
-                },750));
+                },timeoutLength + 100);
             });
-            boardSerial.on('close',function(){
-                reject('Serial Port Closed!');
+            boardSerial.on('close',() => {
+                if (this.options.verbose) console.log('Serial Port Closed');
+                //this.emit('close')
```
Comment (teonlamont) [inline openBCIBoard.js]: why is this commented out?

File: openBCIBoard.js [hunk 2]
```diff
@@ -75,60 +94,78 @@ function OpenBCIFactory() {
...
+            var boardSerial;
+            /* istanbul ignore else */
+            if (this.options.simulate) {
+                if (this.options.verbose) console.log('using faux board ' + portName);
+                boardSerial = new openBCISimulator.OpenBCISimulator(portName, {
+                    verbose: this.options.verbose
+                });
+            } else {
+                /* istanbul ignore if */
+                if (this.options.verbose) console.log('using real board ' + portName);
```
Comment (teonlamont) [inline openBCIBoard.js]: so the simulator requires a portname?

File: openBCISample.js [hunk 3]
```diff
@@ -61,6 +60,35 @@ module.exports = {
...
+
+        // start byte
+        packetBuffer[0] = k.OBCIByteStart;
+
+        // sample number
+        packetBuffer[1] = sample.sampleNumber;
+
+        // channel data
+        for (var i = 0; i < k.OBCINumberOfChannelsDefault; i++) {
+            var threeByteBuffer = this.floatTo3ByteBuffer(sample.channelData[i]);
```
Comment (teonlamont) [inline openBCISample.js]: `floatTo3Byte`? is this to go back the other way for the simulator code?

File: openBCISample.js [hunk 4]
```diff
@@ -78,20 +106,80 @@ module.exports = {
...
 
-            var impedanceArray = [];
-            for (var i = 1; i <= k.OBCINumberOfChannelsDefault; i++) {
-                impedanceArray[i] = (sqrt2 * sampleObject.channelData[i]) / k.OBCILeadOffDriveInAmps;
+            if (sampleObject.channelData[index] < 0) {
+                sampleObject.channelData[index] *= -1;
             }
-            sampleObject.impedanceArray = impedanceArray;
-            resolve(sampleObject);
+            var impedance = (sqrt2 * sampleObject.channelData[index]) / k.OBCILeadOffDriveInAmps;
```
Comment (teonlamont) [inline openBCISample.js]: why is this the calculation?

Comment (teonlamont) [inline openBCISample.js]: is `sambleObject.channelData[index]` the output the `impedanceSummarize` below?

File: openBCISimulator.js [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        }, intervalInMS);
+    };
+
+    factory.OpenBCISimulator = OpenBCISimulator;
+
+}
+
+util.inherits(OpenBCISimulatorFactory, EventEmitter);
+
+module.exports = new OpenBCISimulatorFactory();
```
Comment (teonlamont) [inline openBCISimulator.js]: just out of curiosity, does js formatting typically have a blank line at the end? having one is typical of python and I see that this one doesn't have it

File: test/OpenBCIConstants-test.js [hunk 6]
```diff
@@ -636,7 +636,7 @@ describe('OpenBCIConstants', function() {
         //'channel 1, power on, gain 24, inputType normal, bias include, srb2 connect, srb1 dissconnect'
         describe('channel input selection works', function() {
             //this.timeout(5000);
-            it('channel 2', function() {
+            it('channel 2', function(done) {
```
Comment (teonlamont) [inline test/OpenBCIConstants-test.js]: why does it take `done` now as an argument?

File: test/OpenBCISample-test.js [hunk 7]
```diff
@@ -55,9 +56,8 @@ describe('OpenBCISample',function() {
             //console.log(temp);
             var taco = new Buffer([0x81]);
             taco.copy(temp,2);
-            //console.log(temp);
-            var sample = OpenBCISample.convertPacketToSample(sampleBuf);
-            assert(sample.channelData[1], -8323071 * OpenBCISample.scaleFactorChannel);
+            var sample = openBCISample.convertPacketToSample(temp);
+            sample.channelData[0].should.be.approximately(-8323071 * openBCISample.scaleFactorChannel, 0.001);
```
Comment (teonlamont) [inline test/OpenBCISample-test.js]: `should.be.approximately` is built-in for comparison? if so, that's cool. if not, what is that doing?

File: test/openbci-sdk-test.js [hunk 8]
```diff
@@ -248,84 +602,658 @@ describe('openbci-sdk',function() {
...
-                ourBoard.simulatorStop().then(function() {
-                    running = true;
-                    done();
-                },function(err) {
-                    console.log('Error: ' + err);
+
+
+});
+
+describe('#impedanceTesting', function() {
```
Comment (teonlamont) [inline test/openbci-sdk-test.js]: can the describes below be looped over instead of having separate ones per channel? it would make that a bit cleaner especially since everything is the same except the channel number

File: openBCIBoard.js [hunk 9]
```diff
@@ -238,24 +322,110 @@ function OpenBCIFactory() {
         });
     };
 
+    /**
+     * @description Should be used to send data to the board
+     * @param data
+     * @returns {Promise} if signal was able to be sent
+     * @author AJ Keller (@pushtheworldllc)
+     */
+    OpenBCIBoard.prototype._writeAndDrain = function(data) {
```
Comment (teonlamont) [inline openBCIBoard.js]: does serialPort have a built-in writeAndDrain. if so, what makes this one different from that?

**Comment by teonlamont** [comment]
this was such a massive code dump ;) I made it through it.
are you going to add the option for the data to be streamed as both float and binary (byte packets)? if not, what is the fidelity of the data being converted back to the binary? I know you have a test of it but what is the tolerance or the difference from the original binary and the binary to float to binary conversion?

**Comment by teonlamont** [comment]
btw, the git history is fine. it wouldn't be a problem with merging.
my plans for OpenEXP is to create an EEG writer, which takes an n_channels x n_samples and writes it to EDF (loss of resolution from 24-bit to 16-bit) or BDF (24-bit).

File: openBCIBoard.js (hunk 2 above)
Comment (pushtheworldllc) [inline openBCIBoard.js]: You can start it a number of ways... * Sending connect "/dev/tty.openBCISimulator"* Setting `simulate` true on constructor call* calling `simulatorStart()`

AJ Keller
Founder 
Push The World
843-471-0039




On Fri, Feb 5, 2016 at 9:58 AM -0800, "Teon L Brooks" <notifications@github.com> wrote:












In openBCIBoard.js:
> -            var boardSerial = new serialPort.SerialPort(portName, {
> -                baudRate: this.options.baudRate
> -            },(err) => {
> -                reject(err);
> -            });
> +            // If we are simulating, set boardSerial to fake name
> +            var boardSerial;
> +            /* istanbul ignore else */
> +            if (this.options.simulate) {
> +                if (this.options.verbose) console.log('using faux board ' + portName);
> +                boardSerial = new openBCISimulator.OpenBCISimulator(portName, {
> +                    verbose: this.options.verbose
> +                });
> +            } else {
> +                /* istanbul ignore if */
> +                if (this.options.verbose) console.log('using real board ' + portName);



so the simulator requires a portname?

File: openBCISample.js (hunk 3 above)
Comment (pushtheworldllc) [inline openBCISample.js]: Yup!!

AJ Keller
Founder 
Push The World
843-471-0039




On Fri, Feb 5, 2016 at 1:44 PM -0800, "Teon L Brooks" <notifications@github.com> wrote:












In openBCISample.js:
> @@ -61,6 +60,35 @@ module.exports = {
>              stopByte: dataBuf[numberOfBytes - 1] // byte
>          }
>      },
> +    convertSampleToPacket: function(sample) {
> +        var packetBuffer = new Buffer(k.OBCIPacketSize);
> +        packetBuffer.fill(0);
> +
> +        // start byte
> +        packetBuffer[0] = k.OBCIByteStart;
> +
> +        // sample number
> +        packetBuffer[1] = sample.sampleNumber;
> +
> +        // channel data
> +        for (var i = 0; i < k.OBCINumberOfChannelsDefault; i++) {
> +            var threeByteBuffer = this.floatTo3ByteBuffer(sample.channelData[i]);



floatTo3Byte? is this to go back the other way for the simulator code?

File: openBCISample.js (hunk 4 above)
Comment (pushtheworldllc) [inline openBCISample.js]: This is the way chip did it on the Processing GUI... Checkout his blog post on eeghacker 

AJ Keller
Founder 
Push The World
843-471-0039




On Fri, Feb 5, 2016 at 1:49 PM -0800, "Teon L Brooks" <notifications@github.com> wrote:












In openBCISample.js:
>              }
> -            sampleObject.impedanceArray = impedanceArray;
> -            resolve(sampleObject);
> +            var impedance = (sqrt2 * sampleObject.channelData[index]) / k.OBCILeadOffDriveInAmps;



why is this the calculation?

File: openBCIBoard.js (hunk 9 above)
Comment (pushtheworldllc) [inline openBCIBoard.js]: No they do not!

File: openBCISample.js (hunk 4 above)
Comment (pushtheworldllc) [inline openBCISample.js]: Could you rephrase the second line note please?

File: openBCISimulator.js (hunk 5 above)
Comment (pushtheworldllc) [inline openBCISimulator.js]: I'm really not sure! I just do it for readability

File: test/OpenBCISample-test.js (hunk 7 above)
Comment (pushtheworldllc) [inline test/OpenBCISample-test.js]: Nope! It is an extension of the chai testing framework. But it makes testing much more natural instead of a bunch of asserts and such

File: test/openbci-sdk-test.js (hunk 8 above)
Comment (pushtheworldllc) [inline test/openbci-sdk-test.js]: I tried to do it but everything sort of broke on me and I just went with the try to make it work mentality hahah. I read that you could if you like set it up in another function or something weird like that but I couldn't figure it out.

File: openBCIBoard.js (hunk 1 above)
Comment (pushtheworldllc) [inline openBCIBoard.js]: ok i uncommented it and added it to the ReadMe

File: test/OpenBCIConstants-test.js (hunk 6 above)
Comment (pushtheworldllc) [inline test/OpenBCIConstants-test.js]: the mocha framework tests async functions with the done argument... you call it when the thing you are testing finishes, it's essentially tells the `it` to wait till is gets called.

**Comment by pushtheworldllc** [comment]
not sure what happened!

File: openBCISample.js (hunk 4 above)
Comment (teonlamont) [inline openBCISample.js]: you have `sambleObject.channelData[index]` in the impedance calculation, I was wondering where this gets defined. I didn't see it anywhere. I am assuming these are the output values from sending the impedance command to the channel.

File: test/OpenBCIConstants-test.js (hunk 6 above)
Comment (teonlamont) [inline test/OpenBCIConstants-test.js]: got it

File: openBCISample.js (hunk 4 above)
Comment (pushtheworldllc) [inline openBCISample.js]: ok so you can sampleObject is sent into this method as input. This method is called in `._processBytes()` every time a 33-byte raw packets is converted to a sample object AND we are calculating impedance.

File: test/openbci-sdk-test.js (hunk 8 above)
Comment (teonlamont) [inline test/openbci-sdk-test.js]: ok, cool.

File: test/OpenBCISample-test.js (hunk 7 above)
Comment (teonlamont) [inline test/OpenBCISample-test.js]: did you look at how different the comparison is between the two

Comment (pushtheworldllc) [inline test/OpenBCISample-test.js]: A comparison between chai should and assert? Or what the differences are between what I'm testing? The way this test is currently if something broke withr the conversion we would know here

Comment (teonlamont) [inline test/OpenBCISample-test.js]: no, sorry, the actual data values. you could do mean(original - convertedOriginal) to see what size of a difference there is. we know that is not perfect because of precision, otherwise we would be able to assert_equal.

**Comment by teonlamont** [comment]
LGTM, ready for me to merge it?

**Comment by teonlamont** [comment]
thanks @pushtheworldllc!

**Comment by pushtheworldllc** [comment]
Thanks @teonlamont for your code review and insights!
